ha/ha.go

```diff
@@ -207,151 +230,130 @@ func (h *HA) setAndInsertEnvironment(env *Environment) error {
+		err := h.checkConflictingInstances(tx)
 		if err != nil {
-			h.logger.Errorf("Failed to insert/update instance: %v", err)
-			h.super.ChErr <- errors.New("failed to insert/update instance")
-			return
+			return err
 		}
 
-		h.isActive = true
-	} else {
-		h.logger.Info("Other instance is active.")
-		h.isActive = false
-		h.lastEventId = "0-0"
-	}
-}
-
-func (h *HA) runHA(chEnv chan *Environment) {
-	select {
-	case env := <-chEnv:
-		if bytes.Compare(env.ID, h.super.EnvId) != 0 {
-			h.logger.Error("Received environment is not the one we expected. Panic.")
-			h.super.ChErr <- errors.New("received unexpected environment")
-			return
+		foundActive, activeId, err := h.getActiveInstance(tx)
+		if err != nil {
+			return err
 		}
 
-		h.heartbeatTimer.Reset(time.Second * 15)
-		previous := h.lastHeartbeat
-		h.lastHeartbeat = utils.TimeToMillisecs(time.Now())
-
-		if h.lastHeartbeat-previous < 10*1000 && h.isActive {
-			err := h.updateOwnInstance(env)
-
-			if err != nil {
-				h.logger.Errorf("Failed to update instance: %v", err)
-				h.super.ChErr <- errors.New("failed to update instance")
-				return
+		lastIcinga2HeartbeatAge := utils.TimeToMillisecs(time.Now()) - h.lastHeartbeat
+		lastIcinga2HeartbeatValid := lastIcinga2HeartbeatAge < heartbeatValidMillisecs
+
+		if foundActive {
+			if bytes.Equal(activeId, h.instanceId[:]) {
+				if lastIcinga2HeartbeatValid {
+					// We are active according to the DB and have a valid heartbeat, keep it that way.
+					newState = StateActive
+				} else {
+					// We are active according to the DB but our heartbeat is no longer valid, give up active state.
+					newState = StateAllInactive
+				}
+			} else {
+				// someone else is active, remain passive
+				newState = StateOtherActive
 			}
 		} else {
-			_, they, beat, err := h.getInstance()
-			if err != nil {
-				h.logger.Errorf("Failed to fetch instance: %v", err)
-				h.super.ChErr <- errors.New("failed to fetch instance")
-				return
-			}
-			if they == h.uid {
-				h.logger.Debug("We are active.")
-				if !h.isActive {
-					h.logger.Info("Icinga 2 sent heartbeat. Starting sync")
-					h.isActive = true
-				}
-
-				if err := h.updateOwnInstance(env); err != nil {
-					h.logger.Errorf("Failed to update instance: %v", err)
-					h.super.ChErr <- errors.New("failed to update instance")
-					return
-				}
-			} else if h.lastHeartbeat-beat > 15*1000 {
-				h.logger.Info("Taking over.")
-				if err := h.takeOverInstance(env); err != nil {
-					h.logger.Errorf("Failed to update instance: %v", err)
-					h.super.ChErr <- errors.New("failed to update instance")
-				}
-				h.isActive = true
+			if lastIcinga2HeartbeatValid {
+				// No active instance but we are able to take over. Do so, unless this transaction fails.
+				h.logger.Info("No active instance, trying to take over.")
+				newState = StateActive
 			} else {
-				h.logger.Debug("Other instance is active.")
+				newState = StateAllInactive
```

Comment:
This is to only attempt a takeover if we are actively receiving data from Icinga 2. Required to prevent issues like #233. Should get an additional comment explaining this.